### ISSUE TITLE
Add the ability to have replay produce a C++ program

### ISSUE THREAD

**Issue by jrmuizel** [apitrace/apitrace#351]
This would be enormously valuable for reducing test cases/editing traces.

**Comment by jrfonseca** [comment]
Are you volunteering? ;-)

Now seriously, I agree it would be interesting, I'm fine in principle if somebody is willing to do it, and it wouldn't even be hard for a person familiar with Python to take `scripts/unpickle.py` start hacking something that would consume a .trace and dump C++, leveraging the `specs/*.py` (to know what are the output variables, and handles, etc).

But at the end of the day, to get this working reliably with all sort of traces, it would basically need to replicate all functionality of gl/d3dretrace, and that's *a lot* of functionality.

So even if I agree this would be nice, I just don't have the time to provide and maintain duplicate and complex solutions to fundamentally the same problem.  Instead, I rather keep focus on improving only one.

In short, even they don't provide ultimate flexibility, I'd rather keep focusing on gl/d3dretrace for retracing, and qapitrace for editing.

But I might also provide CLI-based trace editor (ie, dumps all calls to a text file, all blobs to files, then allow one to edit the dump with an ordinary text editor, and recompile the trace back.)

**Comment by jrfonseca** [comment]
For the record, @juhapekka [annouced](http://lists.freedesktop.org/archives/apitrace/2015-September/001121.html) he started a [prototype](https://github.com/juhapekka/apitrace-1/tree/c_source_code_writer) this feature.

**Comment by jrmuizel** [comment]
I also put something together that worked for building a reduced test case for a driver bug:
https://github.com/jrmuizel/apitrace/tree/tocpp

**Comment by jrfonseca** [comment]
Very well.  This functionality will never be as reliable as glretrace, but if folks find it useful to have a script which automates some of the work of generating a C++, but then requires manual editing, I don't object.   All I request is a clear disclaimer stating YMMV, bugs are likely, and manual editing is to be expected.

**Comment by juhapekka** [comment]
@jrmuizel @jrfonseca I have this tree which has not been touched for several years https://github.com/juhapekka/apitracepy I remember getting quite complex traces working back then but there was still holes in the implementation. Past couple of years there were some people using this for something and I guess it did work for them. I have vague memory there was some issue osx vs linux in python which causing some onliner runtime python issue on one against other. cwriter.py is the startup script.
